### PR TITLE
fix bug in deflate with windowBits == 8

### PR DESCRIPTION
**Pull request by MylesBorins**
This patch comes from the optipng project and is licensesd under the
zlib license. I am unsure the original author but it may be @ctruta

    This patch applies to zlib-1.1.3. It fixes a problem with 256-byte
    sliding windows, that exists in all the zlib versions up to 1.1.3,
    and it is proposed to be included in the future zlib-1.1.4 release
    The problem occurs because MIN_LOOKAHEAD in deflate.c is usually
    bigger than 256, and this confuses the longest_match() function.

    The validity of the patch is enabled by this spec correction. The
    particular implementation of the reference library limits the
    distance codes to 2^windowBits - MIN_LOOKAHEAD.
    When windowBits =9, all the distance codes are no bigger than
    250, so the CINFO flag in the zlib stream can safely be 0.

    The patch also extends the functionality of deflateInit(), so that
    other (smaller) window sizes are also accepted. This way, the
    library becomes more extensible, allowing future additions of
    newer strategies optimized for speed and/or low memory
    requirements, such as the Z_RLE strategy. If a certain value of
    windowBits is not directly supported by the implementation, it can
    be adjusted to the nearest supported value: valid zlib streams are
    still produced.

Refs: http://optipng.sourceforge.net/pngtech/zlib-256win-bug.html

### CODE REVIEW

File: deflate.c [hunk 1]
```diff
@@ -296,11 +296,10 @@ int ZEXPORT deflateInit2_(strm, level, method, windowBits, memLevel, strategy,
     }
 #endif
     if (memLevel < 1 || memLevel > MAX_MEM_LEVEL || method != Z_DEFLATED ||
-        windowBits < 8 || windowBits > 15 || level < 0 || level > 9 ||
```
Comment (jcoglan) [inline deflate.c]: @MylesBorins What's the reason the check for `windowBits < 8` has been removed here?

Comment (MylesBorins) [inline deflate.c]: Because it is no longer an error

Comment (jcoglan) [inline deflate.c]: I'm probably missing something but I thought that bug was just in `windowBits=8`. Does fixing the 256-byte window bug mean all values of `windowBits` from 0 to 15 are now valid?

Comment (MylesBorins) [inline deflate.c]: in retrospect that is what is in the original patch mentioned in the refernce

**Comment by timofonic** [comment]
What about merging it to @Dead2 [zlib-ng](/Dead2/zlib-ng)?

File: deflate.c [hunk 2]
```diff
@@ -309,7 +308,11 @@ int ZEXPORT deflateInit2_(strm, level, method, windowBits, memLevel, strategy,
 
     s->wrap = wrap;
     s->gzhead = Z_NULL;
-    s->w_bits = (uInt)windowBits;
+#if MIN_LOOKAHEAD < 256
+    s->w_bits = ((uInt)windowBits >= 8) ? (uInt)windowBits : 8;
+#else
+    s->w_bits = ((uInt)windowBits >= 9) ? (uInt)windowBits : 9;
+#endif
     s->w_size = 1 << s->w_bits;
```
Comment (peter-gribanov) [inline deflate.c]: As far as i can understand, a 256 byte window is not supported at all #171. Therefore, it is more correct to write like this:
```diff
-    s->w_bits = (uInt)windowBits;
+    s->w_bits = ((uInt)windowBits >= 9) ? (uInt)windowBits : 9;
```

File: deflate.c [hunk 3]
```diff
@@ -813,7 +816,12 @@ int ZEXPORT deflate (strm, flush)
     /* Write the header */
     if (s->status == INIT_STATE) {
         /* zlib header */
+#if MIN_LOOKAHEAD < 256
         uInt header = (Z_DEFLATED + ((s->w_bits-8)<<4)) << 8;
+#else
+        uInt optimized_cinfo = (s->w_bits > 9) ? s->w_bits - 8 : 0;
+        uInt header = (Z_DEFLATED + (optimized_cinfo<<4)) << 8;
+#endif
```
Comment (peter-gribanov) [inline deflate.c]: As far as i can understand, a 256 byte window is not supported at all #171. Therefore, it is more correct to write like this:
```diff
-        uInt header = (Z_DEFLATED + ((s->w_bits-8)<<4)) << 8;
+        uInt optimized_cinfo = (s->w_bits > 9) ? s->w_bits - 8 : 0;
+        uInt header = (Z_DEFLATED + (optimized_cinfo<<4)) << 8;

```

**Pull request closed without merging**